fal/controllers/anime_stats.py

```diff
@@ -0,0 +1,77 @@
+from __future__ import annotations
+
+import configparser
+from typing import TYPE_CHECKING, Dict, Union, List, cast
+
+import jikanpy
+
+from fal.clients.mfalncfm_main import session_scope
+from fal.models import AnimeWeeklyStat, Season, Anime
+
+if TYPE_CHECKING:
+    from sqlalchemy.orm import Session
+
+config = configparser.ConfigParser()
+config.read("config.ini")
+
+
+def get_forum_posts(anime: Anime) -> int:
+    '''Not yet implemented'''
+    return 0
+
+
+def get_anime_stats_from_jikan(anime: Anime) -> Dict[str, Union[int, float]]:
+    '''
+    Makes requests to Jikan given an anime id
+    and returns a Dict containing all the information needed to
+    populate an AnimeWeeklyStat object
+    '''
+
+    jikan = jikanpy.Jikan()
+    general_anime_info = jikan.anime(anime.id)
+    jikan_anime_stats = jikan.anime(anime.id, extension='stats')
+
+    return {
+        'watching': jikan_anime_stats['watching'],
+        'completed': jikan_anime_stats['completed'],
+        'dropped': jikan_anime_stats['dropped'],
+        'score': general_anime_info['score'],
+        'favorites': general_anime_info['favorites'],
+        'forum_posts': get_forum_posts(anime)
+    }
+
+
+def populate_anime_weekly_stats() -> None:
+    '''
+    Populates the AnimeWeeklyStat table with a row for each anime
+    using data from Jikan.
+    '''
+
+    season_of_year = config.get("season info", "season").lower()
+    year = config.getint("season info", "year")
+    week = config.getint("weekly info", "current-week")
+
+    with session_scope() as session:
+        anime_list = Season.get_season_from_database(
+            season_of_year, year, session).anime
+
+        # casting until update in sqlalchemy-stubs
+        for anime in cast(List[Anime], anime_list):
+            stat_data = get_anime_stats_from_jikan(anime)
+            stat_data.update({
+                'week': week,
+                'anime_id': anime.id
+            })
+
+            query = session.query(AnimeWeeklyStat). \
+                filter(AnimeWeeklyStat.week == week,
+                       AnimeWeeklyStat.anime_id == anime.id)
+            anime_weekly_stat = query.one_or_none()
+
+            if not anime_weekly_stat:
+                anime_weekly_stat = AnimeWeeklyStat()
+
+            for key, value in stat_data.items():
+                setattr(anime_weekly_stat, key, value)
+
+            session.add(anime_weekly_stat)
```

Comment:
I was reading up on how to do an INSERT_OR_UPDATE in SQLAlchemy and apparently [merging](https://docs.sqlalchemy.org/en/latest/orm/session_state_management.html#merging) is what is used. I'm not 100% sure what add does if the object already exists, which is why I brought up this alternative, but if add works, then we can stick with it.

Reply:
Actually I'm pretty sure that's a bug that I haven't tested for. Awesome find with merging, we should definitely use that moving forward!